### PR TITLE
[TF:TRT] Create execution context with device memory if shape output is present in TRT 7

### PR DESCRIPTION
**Pull request by tfeher**
Workaround for TRT 7 to handle shape outputs.

Previously, we use createExecutionContextWithoutDeviceMemory to create IExecutionContext, so that TF-TRT can manage the needed device memory to support the execution using the IExecutionContext. This triggers a bug in TRT 7. To workaround the bug, we switch to use createExecutionContext for creating IExecutionContext when the TRTEngine has int32 output.

Add two test cases.

Tracker #45481

Tagging @bixia1 for review.

### CODE REVIEW

**Comment by bixia1** [comment]
Looks good to me. I modified the PR description with more details. Please check.

File: tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc [hunk 1]
```diff
@@ -37,9 +37,24 @@ namespace tensorrt {
...
+  bool has_int32_output = false;
+  for (int i = 0; i < cuda_engine->getNbBindings(); i++) {
+    if (!cuda_engine->bindingIsInput(i) &&
+        cuda_engine->getBindingDataType(i) == nvinfer1::DataType::kINT32) {
+      has_int32_output = true;
+      break;
+    }
+  }
+  if (!IS_TRT_VERSION_GE(8, 0, 0, 0) && has_int32_output) {
+    // Workaround for TRT 7 bug with shape output (nvbugs/3390469).
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc:49]: Let's turn this comment into a TODO:
// TODO(nvbugs/3390469): Remove this workaround when the bug is fixed.

Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc]: Updated.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review, I have updated the comment. Also checked the modified PR description, it looks fine.

**Comment by christopherbate** [comment]
Based on the code then it's not the fact that there is a shape tensor, but that it is int32 data type. That's not the same thing, right? Because shape tensor has additional qualifications of being 0-d or 1-d.

Does this same behavior occur when index output of "topK" operation is set as output?

Is there no op that we can insert at the output to force TRT into correct behavior?

**Comment by tfeher** [comment]
Thanks @christopherbate for your comment! You are right, we could and should be more specific when checking the output tensor. I have added checks to ensure that it is 0-d or 1-d.

> Does this same behavior occur when index output of "topK" operation is set as output?

The error happens at an assertion of a shape tensor size: "Assertion failed: size == shapeTensorVolume(out.extent)"
I expect that the problem would only appear with int32 outputs that are shape tensors.

> Is there no op that we can insert at the output to force TRT into correct behavior?

I am not aware of any.

**Comment by bixia1** [comment]
I am not sure whether we can/should add check for 0-d or 1-d to filter the cases. It depends on the situation of the bug. For example, if we have this in the network:
     shapeOp(4-d) -> reshape(2x2)
Do we need to allocate context with GPU memory to make this work?

**Comment by tfeher** [comment]
@bixia1, you are right. I have checked it, and the error happens if I insert the reshape. I will revert the change, which filters for the 0 and 1d cases. 

I will also add two unit tests that would trigger the bug. To add the tests, I will need to rebase this PR to #52181.

**Comment by tfeher** [comment]
Rebased to avoid conflicts. Ready to proceed @bixia1

**Review by bixia1 — COMMENTED** [review]
Looks good to me, thanks for working on this!
Just two minor comments and it should be good for merging after this.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia for the review, I have addressed the issues.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**